File: crates/moredakka-core/src/doctor.rs
use crate::config::{load_config, AppConfig};
use crate::util::{load_local_env, run_command};
use serde::Serialize;
use std::collections::HashSet;
use std::fs;
use std::path::Path;

#[derive(Clone, Debug, Serialize)]
pub struct DoctorCheck {
    pub name: String,
    pub status: String,
    pub summary: String,
    pub detail: String,
    pub fix: String,
}

#[derive(Clone, Debug, Serialize)]
pub struct DoctorReport {
    pub ok: bool,
    pub cwd: String,
    pub config_path: Option<String>,
    pub checks: Vec<DoctorCheck>,
}

fn check_python() -> DoctorCheck {
    let probe = [
        "/opt/homebrew/bin/python3.12",
        "/opt/homebrew/bin/python3.11",
        "/usr/local/bin/python3.12",
        "/usr/local/bin/python3.11",
        "python3.12",
        "python3.11",
        "python3",
    ]
        .into_iter()
        .find_map(|bin| {
            run_command(
                Path::new("."),
                bin,
                &["-c", "import sys; print(f'{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}')"],
            )
            .map(|version| (bin.to_string(), version))
        });
    match probe {
        Some((bin, version)) => {
            let parts: Vec<u32> = version.split('.').filter_map(|p| p.parse().ok()).collect();
            if parts.len() >= 2 && (parts[0], parts[1]) >= (3, 11) {
                DoctorCheck {
                    name: "python".into(),
                    status: "pass".into(),
                    summary: "Python version is supported.".into(),
                    detail: format!("{bin} {version}"),
                    fix: String::new(),
                }
            } else {
                DoctorCheck {
                    name: "python".into(),
                    status: "fail".into(),
                    summary: "Python 3.11+ is required.".into(),
                    detail: format!("{bin} {version}"),
                    fix: "Use bin/moredakka or a Python 3.11+ interpreter.".into(),
                }
            }
        }
        None => DoctorCheck {
            name: "python".into(),
            status: "warn".into(),
            summary: "Python version could not be checked.".into(),
            detail: String::new(),
            fix: String::new(),
        },
    }
}

fn check_git(cwd: &Path) -> DoctorCheck {
    match run_command(cwd, "git", &["--version"]) {
        Some(version) => DoctorCheck {
            name: "git".into(),
            status: "pass".into(),
            summary: "git is available.".into(),
            detail: version,
            fix: String::new(),
        },
        None => DoctorCheck {
            name: "git".into(),
            status: "fail".into(),
            summary: "git is not available on PATH.".into(),
            detail: String::new(),
            fix: "Install git and ensure it is on PATH.".into(),
        },
    }
}

fn repo_checks(cwd: &Path, base_ref: &str, git_ok: bool) -> (DoctorCheck, DoctorCheck) {
    if !git_ok {
        return (
            DoctorCheck {
                name: "repo".into(),
                status: "warn".into(),
                summary: "Repository checks skipped because git is unavailable.".into(),
                detail: String::new(),
                fix: String::new(),
            },
            DoctorCheck {
                name: "base_ref".into(),
                status: "warn".into(),
                summary: "Base ref check skipped because git is unavailable.".into(),
                detail: String::new(),
                fix: String::new(),
            },
        );
    }
    let repo_root = run_command(cwd, "git", &["rev-parse", "--show-toplevel"]);
    let Some(repo_root) = repo_root else {
        return (
            DoctorCheck {
                name: "repo".into(),
                status: "warn".into(),
                summary: "Current directory is not inside a git repo.".into(),
                detail: String::new(),
                fix: "Run moredakka inside a git repo for review-oriented workflows.".into(),
            },
            DoctorCheck {
                name: "base_ref".into(),
                status: "warn".into(),
                summary: "Base ref not checked outside a git repo.".into(),
                detail: String::new(),
                fix: String::new(),
            },
        );
    };
    if run_command(cwd, "git", &["rev-parse", "--verify", base_ref]).is_none() {
        return (
            DoctorCheck {
                name: "repo".into(),
                status: "pass".into(),
                summary: "Current directory is inside a git repo.".into(),
                detail: repo_root,
                fix: String::new(),
            },
            DoctorCheck {
                name: "base_ref".into(),
                status: "fail".into(),
                summary: "Configured base ref does not resolve locally.".into(),
                detail: base_ref.into(),
                fix: "Fetch the ref, set defaults.base_ref, or pass --base-ref explicitly.".into(),
            },
        );
    }
    (
        DoctorCheck {
            name: "repo".into(),
            status: "pass".into(),
            summary: "Current directory is inside a git repo.".into(),
            detail: repo_root,
            fix: String::new(),
        },
        DoctorCheck {
            name: "base_ref".into(),
            status: "pass".into(),
            summary: "Base ref resolves locally.".into(),
            detail: base_ref.into(),
            fix: String::new(),
        },
    )
}

fn cache_check(cwd: &Path, cache_dir: &str) -> DoctorCheck {
    let resolved = if Path::new(cache_dir).is_absolute() {
        Path::new(cache_dir).to_path_buf()
    } else {
        cwd.join(cache_dir)
    };
    match fs::create_dir_all(&resolved).and_then(|_| {
        let probe = resolved.join(".doctor-write-test");
        fs::write(&probe, "ok\n")?;
        fs::remove_file(&probe)
    }) {
        Ok(_) => DoctorCheck {
            name: "cache_dir".into(),
            status: "pass".into(),
            summary: "Cache directory is writable.".into(),
            detail: resolved.display().to_string(),
            fix: String::new(),
        },
        Err(err) => DoctorCheck {
            name: "cache_dir".into(),
            status: "fail".into(),
            summary: "Cache directory is not writable.".into(),
            detail: err.to_string(),
            fix: "Set defaults.cache_dir to a writable path.".into(),
        },
    }
}

fn provider_checks(
    config: &AppConfig,
    env: &std::collections::HashMap<String, String>,
) -> Vec<DoctorCheck> {
    let active: HashSet<String> = config.roles.values().map(|r| r.provider.clone()).collect();
    config
        .providers
        .values()
        .map(|provider| {
            let missing_env = env
                .get(&provider.api_key_env)
                .map(|v| v.is_empty())
                .unwrap_or(true);
            if !missing_env {
                DoctorCheck {
                    name: format!("provider:{}", provider.name),
                    status: "pass".into(),
                    summary: format!("{} is ready.", provider.name),
                    detail: format!("kind={} env={}", provider.kind, provider.api_key_env),
                    fix: String::new(),
                }
            } else {
                DoctorCheck {
                    name: format!("provider:{}", provider.name),
                    status: if active.contains(&provider.name) {
                        "fail".into()
                    } else {
                        "warn".into()
                    },
                    summary: format!("{} is not fully ready.", provider.name),
                    detail: format!("missing env {}", provider.api_key_env),
                    fix: format!("export {}", provider.api_key_env),
                }
            }
        })
        .collect()
}

fn roster_diversity(config: &AppConfig) -> DoctorCheck {
    let models: HashSet<String> = config
        .roles
        .values()
        .filter_map(|r| config.providers.get(&r.provider).map(|p| p.model.clone()))
        .collect();
    if models.len() >= 2 {
        DoctorCheck {
            name: "roster_diversity".into(),
            status: "pass".into(),
            summary: "Active role roster uses multiple model families.".into(),
            detail: models.into_iter().collect::<Vec<_>>().join(", "),
            fix: String::new(),
        }
    } else {
        DoctorCheck {
            name: "roster_diversity".into(),
            status: "warn".into(),
            summary: "Active role roster collapses to one model.".into(),
            detail: models.into_iter().next().unwrap_or_default(),
            fix: "Map at least one critical role to a contrast model or provider.".into(),
        }
    }
}

pub fn run_doctor(cwd: &Path) -> anyhow::Result<DoctorReport> {
    let env = load_local_env(cwd);
    let (config, config_path) = load_config(cwd, None)?;
    let mut checks = Vec::new();
    checks.push(check_python());
    let git = check_git(cwd);
    let git_ok = git.status == "pass";
    checks.push(git);
    checks.push(DoctorCheck {
        name: "config".into(),
        status: "pass".into(),
        summary: "Configuration loaded.".into(),
        detail: config_path
            .as_ref()
            .map(|p| p.display().to_string())
            .unwrap_or_else(|| "using built-in defaults".into()),
        fix: String::new(),
    });
    let (repo, base_ref) = repo_checks(cwd, &config.defaults.base_ref, git_ok);
    checks.push(repo);
    checks.push(base_ref);
    checks.push(cache_check(cwd, &config.defaults.cache_dir));
    checks.extend(provider_checks(&config, &env));
    checks.push(roster_diversity(&config));
    let ok = checks.iter().all(|c| c.status != "fail");
    Ok(DoctorReport {
        ok,
        cwd: cwd.display().to_string(),
        config_path: config_path.map(|p| p.display().to_string()),
        checks,
    })
}

pub fn render_markdown(report: &DoctorReport) -> String {
    let mut lines = vec![
        "# moredakka doctor".to_string(),
        String::new(),
        format!("overall: {}", if report.ok { "PASS" } else { "FAIL" }),
        format!("cwd: {}", report.cwd),
        format!(
            "config: {}",
            report
                .config_path
                .clone()
                .unwrap_or_else(|| "built-in defaults".into())
        ),
        String::new(),
        "## checks".to_string(),
    ];
    for check in &report.checks {
        lines.push(format!(
            "- [{}] {}: {}",
            check.status, check.name, check.summary
        ));
        if !check.detail.is_empty() {
            lines.push(format!("  detail: {}", check.detail));
        }
        if !check.fix.is_empty() {
            lines.push(format!("  fix: {}", check.fix));
        }
    }
    format!("{}\n", lines.join("\n"))
}

File: src/moredakka/doctor.py
from __future__ import annotations

import json
import os
import shutil
import sys
from dataclasses import asdict, dataclass, field
from pathlib import Path
from typing import Callable, Mapping, Sequence

from moredakka.config import AppConfig, _default_config, _find_config_path, load_config
from moredakka.runlog import resolved_run_dir
from moredakka.util import ensure_dir, load_local_env, run_command


@dataclass
class DoctorCheck:
    name: str
    status: str
    summary: str
    detail: str = ""
    fix: str = ""


@dataclass
class DoctorReport:
    ok: bool
    cwd: str
    config_path: str | None
    checks: list[DoctorCheck] = field(default_factory=list)

    def to_dict(self) -> dict[str, object]:
        return {
            "ok": self.ok,
            "cwd": self.cwd,
            "config_path": self.config_path,
            "checks": [asdict(check) for check in self.checks],
        }


def _module_available(name: str) -> bool:
    try:
        __import__(name)
        return True
    except Exception:
        return False


def _writable_dir_check(path: Path, *, name: str, fix: str) -> DoctorCheck:
    try:
        ensure_dir(path)
        probe = path / ".doctor-write-test"
        probe.write_text("ok\n", encoding="utf-8")
        probe.unlink()
        return DoctorCheck(
            name=name,
            status="pass",
            summary=("Cache directory is writable." if name == "cache_dir" else "Run artifact directory is writable."),
            detail=str(path),
        )
    except OSError as exc:
        return DoctorCheck(
            name=name,
            status="fail",
            summary=("Cache directory is not writable." if name == "cache_dir" else "Run artifact directory is not writable."),
            detail=f"{path}: {exc}",
            fix=fix,
        )


def _provider_check(
    *,
    provider_name: str,
    provider_kind: str,
    env_var: str,
    env: Mapping[str, str],
    active: bool,
    module_available: Callable[[str], bool],
) -> DoctorCheck:
    module_name = {
        "openai": "openai",
        "gemini": "google.genai",
        "openrouter": "openai",
    }[provider_kind]
    sdk_ok = module_available(module_name)
    key_ok = bool(env.get(env_var))

    issues: list[str] = []
    if not sdk_ok:
        issues.append(f"missing SDK module {module_name}")
    if not key_ok:
        issues.append(f"missing env {env_var}")

    if not issues:
        return DoctorCheck(
            name=f"provider:{provider_name}",
            status="pass",
            summary=f"{provider_name} is ready.",
            detail=f"kind={provider_kind} env={env_var}",
        )

    status = "fail" if active else "warn"
    fix_parts: list[str] = []
    if not sdk_ok:
        package = "google-genai" if provider_kind == "gemini" else "openai"
        fix_parts.append(f"install {package}")
    if not key_ok:
        fix_parts.append(f"export {env_var}")
    role_text = "active in the current role map" if active else "configured but not active by default"
    return DoctorCheck(
        name=f"provider:{provider_name}",
        status=status,
        summary=f"{provider_name} is not fully ready.",
        detail=f"{role_text}; " + ", ".join(issues),
        fix="; ".join(fix_parts),
    )


def _roster_diversity_check(config: AppConfig) -> DoctorCheck:
    active_models = {
        config.providers[role.provider].model
        for role in config.roles.values()
        if role.provider in config.providers
    }
    if len(active_models) >= 2:
        return DoctorCheck(
            name="roster_diversity",
            status="pass",
            summary="Active role roster uses multiple model families.",
            detail=", ".join(sorted(active_models)),
        )
    only = next(iter(active_models), "(none)")
    return DoctorCheck(
        name="roster_diversity",
        status="warn",
        summary="Active role roster collapses to one model.",
        detail=only,
        fix="Map at least one critical role to a contrast model or provider.",
    )


def _git(args: list[str], cwd: Path) -> str:
    result = run_command(["git", *args], cwd=cwd)
    if result.returncode != 0:
        return ""
    return result.stdout.rstrip("\n")


def _repo_check(cwd: Path, *, git_available: bool, base_ref: str) -> tuple[DoctorCheck, DoctorCheck]:
    if not git_available:
        return (
            DoctorCheck(name="repo", status="warn", summary="Repository checks skipped because git is unavailable."),
            DoctorCheck(name="base_ref", status="warn", summary="Base ref check skipped because git is unavailable."),
        )

    repo_root = _git(["rev-parse", "--show-toplevel"], cwd)
    if not repo_root:
        return (
            DoctorCheck(
                name="repo",
                status="warn",
                summary="Current directory is not inside a git repo.",
                fix="Run moredakka inside a git repo for review-oriented workflows.",
            ),
            DoctorCheck(
                name="base_ref",
                status="warn",
                summary="Base ref not checked outside a git repo.",
            ),
        )

    if not _git(["rev-parse", "--verify", base_ref], cwd):
        return (
            DoctorCheck(name="repo", status="pass", summary="Current directory is inside a git repo.", detail=repo_root),
            DoctorCheck(
                name="base_ref",
                status="fail",
                summary="Configured base ref does not resolve locally.",
                detail=base_ref,
                fix="Fetch the ref, set defaults.base_ref, or pass --base-ref explicitly.",
            ),
        )

    return (
        DoctorCheck(name="repo", status="pass", summary="Current directory is inside a git repo.", detail=repo_root),
        DoctorCheck(name="base_ref", status="pass", summary="Base ref resolves locally.", detail=base_ref),
    )


def run_doctor(
    *,
    cwd: Path,
    config_path: str | None = None,
    env: Mapping[str, str] | None = None,
    version_info: Sequence[int] | None = None,
    which: Callable[[str], str | None] | None = None,
    module_available: Callable[[str], bool] | None = None,
) -> DoctorReport:
    env_map = load_local_env(cwd, env=env)
    version = tuple(version_info or sys.version_info)
    which_fn = which or shutil.which
    module_ok = module_available or _module_available
    resolved_cwd = cwd.resolve()

    checks: list[DoctorCheck] = []

    if version >= (3, 11):
        checks.append(
            DoctorCheck(
                name="python",
                status="pass",
                summary="Python version is supported.",
                detail=".".join(str(part) for part in version[:3]),
            )
        )
    else:
        checks.append(
            DoctorCheck(
                name="python",
                status="fail",
                summary="Python 3.11+ is required.",
                detail=".".join(str(part) for part in version[:3]),
                fix="Use bin/moredakka or a Python 3.11+ interpreter.",
            )
        )

    git_path = which_fn("git")
    git_available = bool(git_path)
    if git_path:
        checks.append(
            DoctorCheck(
                name="git",
                status="pass",
                summary="git is available.",
                detail=git_path,
            )
        )
    else:
        checks.append(
            DoctorCheck(
                name="git",
                status="fail",
                summary="git is not available on PATH.",
                fix="Install git and ensure it is on PATH.",
            )
        )

    discovered_config = _find_config_path(config_path, resolved_cwd)
    config = _default_config()
    try:
        config = load_config(cwd=resolved_cwd, explicit_path=config_path)
        checks.append(
            DoctorCheck(
                name="config",
                status="pass",
                summary="Configuration loaded.",
                detail=str(discovered_config) if discovered_config else "using built-in defaults",
            )
        )
    except RuntimeError as exc:
        checks.append(
            DoctorCheck(
                name="config",
                status="fail",
                summary="Configuration could not be loaded.",
                detail=str(exc),
                fix="Fix moredakka.toml or pass --config with a valid file.",
            )
        )

    repo_check, base_ref_check = _repo_check(resolved_cwd, git_available=git_available, base_ref=config.defaults.base_ref)
    checks.extend([repo_check, base_ref_check])

    cache_dir_config = Path(config.defaults.cache_dir).expanduser()
    cache_dir = (cache_dir_config if cache_dir_config.is_absolute() else resolved_cwd / cache_dir_config).resolve()
    checks.append(_writable_dir_check(cache_dir, name="cache_dir", fix="Set defaults.cache_dir to a writable path."))

    run_dir = resolved_run_dir(resolved_cwd, config.defaults.run_dir)
    checks.append(_writable_dir_check(run_dir, name="run_dir", fix="Set defaults.run_dir to a writable path."))

    active_provider_names = {role.provider for role in config.roles.values()}
    for provider_name, provider in config.providers.items():
        checks.append(
            _provider_check(
                provider_name=provider_name,
                provider_kind=provider.kind,
                env_var=provider.api_key_env,
                env=env_map,
                active=provider_name in active_provider_names,
                module_available=module_ok,
            )
        )

    checks.append(_roster_diversity_check(config))

    ok = all(check.status != "fail" for check in checks)
    return DoctorReport(
        ok=ok,
        cwd=str(resolved_cwd),
        config_path=str(discovered_config) if discovered_config else None,
        checks=checks,
    )


def render_doctor_markdown(report: DoctorReport) -> str:
    status_line = "PASS" if report.ok else "FAIL"
    lines = [
        "# moredakka doctor",
        "",
        f"overall: {status_line}",
        f"cwd: {report.cwd}",
        f"config: {report.config_path or 'built-in defaults'}",
        "",
        "## checks",
    ]
    for check in report.checks:
        lines.append(f"- [{check.status}] {check.name}: {check.summary}")
        if check.detail:
            lines.append(f"  detail: {check.detail}")
        if check.fix:
            lines.append(f"  fix: {check.fix}")
    return "\n".join(lines).strip() + "\n"


def render_doctor_json(report: DoctorReport) -> str:
    return json.dumps(report.to_dict(), indent=2, ensure_ascii=False) + "\n"
